[Intel MKL] Use Shard function instead of Eigen device to parallelize Adam kernel.

Zantares opened the pull request:
This could reduce the memory access and get good cache locality for CPU.

modified:
- tensorflow/core/kernels/training_ops.cc
- tensorflow/core/kernels/training_ops.h
- tensorflow/core/kernels/training_ops_gpu.cu.cc

Signed-off-by: Lu Teng teng.lu@intel.com

Zantares committed:
Use Shard function instead of Eigen device to parallelize Adam kernel.
This could reduce the memory access and get good cache locality for CPU.

modified:
- tensorflow/core/kernels/training_ops.cc
- tensorflow/core/kernels/training_ops.h
- tensorflow/core/kernels/training_ops_gpu.cu.cc

Signed-off-by: Lu Teng teng.lu@intel.com

Zantares commented:
Eigen device expression can only update 1 variable once, but Adam needs to update 3 variables and uses 3 expression which would impact the cache locality of CPU. Here use Shard function to replace Eigen device expression.

This patch is tested on NCF model which is in MLPerf 0.5 submission.
It can speed up Adam kernel 15%~30%, and improve 10%~20% overall for the model.

agramesh1 commented:
pinging @ezhulenev  for a review. Thanks.

ezhulenev commented on tensorflow/core/kernels/training_ops.cc:
```diff
@@ -310,21 +310,63 @@ struct ApplyAdamNonCuda {
+      T v;
+      T g;
+
+      if (use_nesterov) {
```
I think both of these loops could be vectorized/simplified with Eigen (see example from redux_functor.h: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/redux_functor.h#L72-L95)

```
    using Input = Eigen::TensorMap<
        Eigen::Tensor<const T, 1, Eigen::RowMajor, Eigen::Index>,
        Eigen::Unaligned>;

    Input m(m_ptr, (begin-end)), v(..), g(..);

    if (use_nesterov) {
      m += (grad - m) * (T(1) - beta1());
      v += (grad.square() - v) * (T(1) - beta2());
      var -= ((grad * (T(1) - beta1()) + beta1() * m) * alpha) / (v.sqrt() + epsilon());
    } else {
       ...
    }
}
```

just remove `.device()` from tensor expressions to evaluate them with default single threaded device

Zantares replied:
> I think both of these loops could be vectorized/simplified with Eigen (see example from redux_functor.h: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/redux_functor.h#L72-L95)
> 
> ```
>     using Input = Eigen::TensorMap<
>         Eigen::Tensor<const T, 1, Eigen::RowMajor, Eigen::Index>,
>         Eigen::Unaligned>;
> 
>     Input m(m_ptr, (begin-end)), v(..), g(..);
> 
>     if (use_nesterov) {
>       m += (grad - m) * (T(1) - beta1());
>       v += (grad.square() - v) * (T(1) - beta2());
>       var -= ((grad * (T(1) - beta1()) + beta1() * m) * alpha) / (v.sqrt() + epsilon());
>     } else {
>        ...
>     }
> }
> ```
> 
> just remove `.device()` from tensor expressions to evaluate them with default single threaded device

Good idea. I also find if use "device.parallelFor" instead "Shard", the function parameter can keep unchanged.

ezhulenev replied:
I realized that memory access pattern might be not as good as in a manual loop, but I'd love to see proofs in micro benchmarks. If there is no performance penalty I'd prefer smaller/simpler to read code.

Zantares replied:
OK, after I refine this code I'll try to make a micro benchmark, please wait 1~2 days, thanks.

Zantares replied:
Done. Please check the new code, thanks!

ezhulenev commented on tensorflow/core/kernels/training_ops.cc:
```diff
@@ -310,21 +310,63 @@ struct ApplyAdamNonCuda {
+               Eigen::TensorOpCost::MulCost<T>() * 6 +
+               Eigen::TensorOpCost::DivCost<T>();
+    auto worker_threads = *(ctx->device()->tensorflow_cpu_worker_threads());
+    Shard(worker_threads.num_threads, worker_threads.workers, length, cost,
```
Please add a comment that because of memory access pattern/locality we are using Shard instead of "regular" tensor expressions.

Zantares replied:
Done.

ezhulenev reviewed (commented):
Could you please also add a benchmark similar to https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/bias_op_test.cc, I'll need it to run performance testing internally.

Zantares committed:
Add comment for Shard function

To get better cache locality, use Shard instead of Eigen expression.

Zantares committed:
Refine code with simple Tensor vectorization form.
Also added a benchmark to test Adam performance.

Zantares commented:
> Could you please also add a benchmark similar to https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/bias_op_test.cc, I'll need it to run performance testing internally.

Hi, @ezhulenev , I've refined the code and added a benchmark https://github.com/tensorflow/tensorflow/pull/26424/files#diff-0b9bd0c5daec98f25d2e15c9b8c0370cR200.

My test result is:

original:
Benchmark             Time(ns) Iterations
-----------------------------------------
BM_SGD/131072            58816      10000        8914.0MB/s 2228.5M items/s
BM_SGD/262144           107837       6335        9723.7MB/s 2430.9M items/s
BM_Adagrad/131072       136092       5126        3852.5MB/s 963.1M items/s
BM_Adagrad/262144       216470       3093        4844.0MB/s 1211.0M items/s
BM_Momentum/131072      126981       5114        4128.9MB/s 1032.2M items/s
BM_Momentum/262144      206378       3434        5080.9MB/s 1270.2M items/s
BM_Adam/131072/0        194416       3452        2696.7MB/s 674.2M items/s
BM_Adam/262144/0        334504       2110        3134.7MB/s 783.7M items/s
**BM_Adam/16777216/1     9864090        100        6803.4MB/s 1700.8M items/s**
BM_RMSProp/131072       187562       3545        2795.3MB/s 698.8M items/s
BM_RMSProp/262144       334770       2180        3132.2MB/s 783.1M items/s
BM_AddSign/131072       512574       1449        1022.9MB/s 255.7M items/s
BM_AddSign/262144       922186        727        1137.1MB/s 284.3M items/s
BM_PowerSign/131072    2179936        311        240.5MB/s 60.1M items/s
BM_PowerSign/262144    3963514        177        264.6MB/s 66.1M items/s

optimized:
Benchmark             Time(ns) Iterations
-----------------------------------------
BM_SGD/131072            69680       9636        7524.2MB/s 1881.0M items/s
BM_SGD/262144            95855       5395        10939.2MB/s 2734.8M items/s
BM_Adagrad/131072       158376       5181        3310.4MB/s 827.6M items/s
BM_Adagrad/262144       234968       2831        4462.6MB/s 1115.7M items/s
BM_Momentum/131072      118026       5495        4442.1MB/s 1110.5M items/s
BM_Momentum/262144      215430       3169        4867.4MB/s 1216.8M items/s
BM_Adam/131072/0        202429       3486        2590.0MB/s 647.5M items/s
BM_Adam/262144/0        328765       1965        3189.4MB/s 797.4M items/s
**BM_Adam/16777216/1     7393820        100        9076.3MB/s 2269.1M items/s**
BM_RMSProp/131072       187683       3003        2793.5MB/s 698.4M items/s
BM_RMSProp/262144       372268       2006        2816.7MB/s 704.2M items/s
BM_AddSign/131072       648737       1000        808.2MB/s 202.0M items/s
BM_AddSign/262144       876228        764        1196.7MB/s 299.2M items/s
BM_PowerSign/131072    2178264        322        240.7MB/s 60.2M items/s
BM_PowerSign/262144    3908483        174        268.3MB/s 67.1M items/s

---

There may be some data variance between different execution, but you can find the optimized is always better than the original.

Zantares commented:
BTW, the Tensor vectorization form has similar performance with the loop version, I guess maybe Eigen can generate same ASM internally, but I didn't dig too deep.

ezhulenev commented on tensorflow/core/kernels/training_ops.cc:
```diff
@@ -310,21 +311,62 @@ struct ApplyAdamNonCuda {
                   typename TTypes<T>::ConstFlat grad, bool use_nesterov) {
+    // Get params length and check if they can be vectorized by packet size.
+    Index length = var.size();
+    Index size = Eigen::internal::packet_traits<T>::size;
```
`packet_size` is more common name for this variable in Eigen/TF codebase

Zantares replied:
Done.

ezhulenev commented on tensorflow/core/kernels/training_ops.cc:
```diff
@@ -310,21 +311,62 @@ struct ApplyAdamNonCuda {
+      length = length / size;
+    } else {
+      size = 1;
+    }
```
There is no need to divide the input size by the packet size, and do "manual vectorization". If it's desirable to have shard size `(end-begin)` to be a multiple of a packet size, you can pass `block_align` to `parallelFor` (see https://bitbucket.org/eigen/eigen/src/4b28c8008901c6d760f48f26ee2e3423fd8a2b40/unsupported/Eigen/CXX11/src/Tensor/TensorDeviceThreadPool.h#lines-185). \

I think this should work:
```
[](Index index) -> Index { return Eigen::divup(index, packet_size); }
```

Zantares replied:
I got some question when try to use this function, please see my comment below.

ezhulenev commented on tensorflow/core/kernels/training_ops_test.cc:
```diff
@@ -176,23 +176,28 @@ static void Adam(int32 n, Graph** init_g, Graph** train_g) {
 }
 
-static void BM_Adam(int iters, int params) {
+static void BM_Adam(int iters, int params, int is_perf) {
```
`is_perf` -> `is_multi_threaded` ? `perf` is too generic term

Zantares replied:
Done.

ezhulenev commented:
I guess after inlining it all might have been fused into a single loop by compiler. Anyway it's great that there is no performance difference and we can keep simpler code.

Zantares committed:
New small benchmark and excatly var name.

Zantares commented:
When I try to use `block_align` to align shard size, I found the performance was decreased in real model, then I captured the param size from model and made a small benchmark: https://github.com/tensorflow/tensorflow/pull/26424/commits/e4dae32cdae1220a19e431e76520b361643e866b#diff-0b9bd0c5daec98f25d2e15c9b8c0370cR200.

current implementation:
---
BM_Adam/8192/1           82133       8295        **399.0MB/s 99.7M items/s**
BM_Adam/16777216/1     8501990        100        7893.3MB/s 1973.3M items/s

with `block_align`:
---
BM_Adam/8192/1           88312       7707        **371.0MB/s 92.8M items/s**
BM_Adam/16777216/1     8462090        100        7930.5MB/s 1982.6M items/s

---
With "manual vectorization", the small benchmark will get better performance(+10%). It's really confused me, maybe Eigen efficiency model https://bitbucket.org/eigen/eigen/src/4b28c8008901c6d760f48f26ee2e3423fd8a2b40/unsupported/Eigen/CXX11/src/Tensor/TensorDeviceThreadPool.h?fileviewer=file-view-default#TensorDeviceThreadPool.h-188 can't handle small size with block_align well?

---
How I use `block_align`
```
+    // Set a function to align block size to packet size, which can get more
+    // chance to vectorize.
+    auto block_align = [packet_size](Index block_size) -> Index {
+      return Eigen::divup(block_size, packet_size) * packet_size;
+    };
+    d.parallelFor(length, cost, block_align, shard)
```
`block_align` will get the `block_size` computed by Eigen efficiency model, it allows us round up the size with own rule and return it as the new block size. I must increase the block size in `block_align` , or it will trap in an infinity loop.
Base on the result, I prefer to use "manual vectorization" version, how do you think about this situation?

ezhulenev commented:
That's strange. I'll try to reproduce it internally after it will be merged.

ezhulenev commented:
I think the problem is in incorrectly computed cost, and Eigen sharding too much or too less.

Zantares committed:
Fix shard cost and var name.

Zantares commented:
> I think the problem is in incorrectly computed cost, and Eigen sharding too much or too less.

Hi @ezhulenev ,please take a look at the new commit, I fixed an error of computing cost - need to multiply `length` to `compute_cycles`. I also checked the CostModel in Eigen, it already had some estimation on cache so I added the store cost. This should be the last commit if no more review suggestion.

The 'manual vectorization' is still better than `block_align`, I guess because compiler could get more static info from 'manual vectorization' while `block_align` may generate a tail that can't be dived in run-time.
